proposals/lip-0051.md

```diff
@@ -1267,46 +1175,95 @@ unlock(address, moduleID, tokenID, amount):
-            beforeExecuteCCM fails
-        availableBalance(relayerAddress, TOKEN_ID_LSK_MAINCHAIN) += fee
-        escrowAmount(CCM.sendingChainID, LOCAL_ID_LSK) -= fee
+        beforeApplyCCM fails
+
+    feeTokenID = interoperability.getChannel(ccm.sendingChainID).messageFeeTokenID
+    # If the chain is the fee token native chain, unescrow and assign fee to relayer
+    if feeTokenID.chainID == interoperability.getOwnChainAccount().ID:
+        if escrowAmount(ccm.sendingChainID, feeTokenID.localID) < fee:
+            interoperability.terminateChain(ccm.sendingChainID)
+            return False
+
+        escrowAmount(ccm.sendingChainID, feeTokenID.localID) -= fee
+        availableBalance(relayerAddress, canonicalTokenID(feeTokenID)) += fee
+    # Else, just assign the fee to relayer
     else:
-        availableBalance(relayerAddress, TOKEN_ID_LSK) += fee    
+        availableBalance(relayerAddress, feeTokenID) += fee    
+```
+
+#### beforeRecoverCCM
+
+```python
+beforeRecoverCCM(trsSender, ccm):
+    fee = ccm.fee
+    if fee < 0:
+        beforeRecoverCCM fails
+
+    feeTokenID = interoperability.getChannel(ccm.sendingChainID).messageFeeTokenID
+    # If the chain is the fee token native chain, unescrow and assign fee to relayer
+    if feeTokenID.chainID == interoperability.getOwnChainAccount().ID:
+        if escrowAmount(ccm.sendingChainID, feeTokenID.localID) < fee:
+            interoperability.terminateChain(ccm.sendingChainID)
+            return False
+
+        escrowAmount(ccm.sendingChainID, feeTokenID.localID) -= fee
+        availableBalance(trsSender, canonicalTokenID(feeTokenID)) += fee
 ```
 
 #### beforeSendCCM
 
 ```python
-beforeSendCCM(payFromAddress, CCM):
-    fee = CCM.fee
+beforeSendCCM(payFromAddress, ccm):
+    fee = ccm.fee
     if fee < 0:
         beforeSendCCM fails
 
-    if interoperability.getOwnChainAccount().ID == MAINCHAIN_ID:
-        # If this chain is the mainchain, escrow the fee
-        if availableBalance(payFromAddress, TOKEN_ID_LSK_MAINCHAIN) < fee:
+    feeTokenID = interoperability.getChannel(ccm.receivingChainID).messageFeeTokenID
+    # If the chain is the fee token native chain, escrow and pay fee
+    if feeTokenID.chainID == interoperability.getOwnChainAccount().ID:
+        if availableBalance(payFromAddress, canonicalTokenID(feeTokenID)) < fee:
             beforeSendCCM fails
-        availableBalance(payFromAddress, TOKEN_ID_LSK_MAINCHAIN) -= fee
-        escrowAmount(CCM.receivingChainID, LOCAL_ID_LSK) += fee
+        availableBalance(payFromAddress, canonicalTokenID(feeTokenID)) -= fee
+        escrowAmount(ccm.receivingChainID, feeTokenID.localID) += fee
+    # Else, pay fee
     else:
-        if availableBalance(payFromAddress, TOKEN_ID_LSK) < fee:
+        if availableBalance(payFromAddress, feeTokenID) < fee:
             beforeSendCCM fails
-        availableBalance(payFromAddress, TOKEN_ID_LSK) -= fee
+        availableBalance(payFromAddress, feeTokenID) -= fee
```

Comment:
I was  first confused why the nested if is necessary and why to once use `canonicalTokenID` and otherwise not. Basically, also the following would work, right? This makes it more clear that you mainly check if the token is native as you then need to escrow.

```
if availableBalance(payFromAddress, canonicalTokenID(feeTokenID)) < fee:
    beforeSendCCM fails
availableBalance(payFromAddress, canonicalTokenID(feeTokenID)) -= fee

if feeTokenID.chainID == interoperability.getOwnChainAccount().ID:
    escrowAmount(ccm.receivingChainID, feeTokenID.localID) += fee
```

Reply:
Yes, this logic would be the same. This nesting was chosen to separate fully the cases where the fee token is native vs the case when it is not.
For us, it was more clear this way.

Reply:
All good, can be resolved.